### ISSUE TITLE
How to get list of packages including dependencies?

### ISSUE THREAD

**Issue by rclai** [meteor/meteor#2853]
`meteor list` shows the list of packages that were manually installed using `meteor add`, is there a way to show the dependencies too like the way `npm list` does?

**Comment by steph643** [comment]
For people landing here, this has been asked on StackOverflow six months ago but not answered:
http://stackoverflow.com/questions/26153603/how-to-show-all-package-dependency-tree

**Comment by rclai** [comment]
This is more of a feature request.

**Comment by trusktr** [comment]
Well, i'm working on a [package](https://github.com/trusktr/rocket-module) that will have such a function to get the dependencies of a given package. When done, I'll split this function (and other useful functions) out into a separate utility package.

**Comment by timothyarmes** [comment]
I've written a utility to provide this feature:

https://www.npmjs.com/package/packagecheck

**Comment by apendua** [comment]
@trusktr Did you eventually succeeded in creating such a utility? I've looked into the problem and from my perspective the biggest blocker is that `meteor show` does not properly extract dependencies from a local package.

**Comment by ferborva** [comment]
Aside from the dependency tree listing, which would be very interesting to have as part of the meteor list tool, I'm missing a license tag on the command. With this I mean it would be very nice to do something like:
`meteor list -l`

and get:
```
  random@1.0.10 - (License Used)
  retry@1.0.9 - ....
  tracker@1.1.1 - ....
  underscore@1.0.10 - ...
  webapp@1.3.12 - ....
```

There is a npm module that does something similar for node dependencies (*nlf*).

**Comment by MichaelJCole** [comment]
From the SO question:
```
for p in `meteor list | grep '^[a-z]' | awk '{ print $1"@"$2 }'`; do echo "$p"; meteor show "$p" | grep -E '^  [a-z]'; echo; done
```

**Comment by abernix** [comment]
While helpful, that command is painfully slow, and on top of it, it fails if the package has any updates available for it or if it's a local package because Meteor appends `*` or `+` to the version number as an indicator.  This results in a lot of these errors:

    reactive-var@1.0.10*
    reactive-var@1.0.10*: not found

Here's a slight modification to the `awk` which fixes that:

    for p in `meteor list | grep '^[a-z]' | awk '{sub(/[+*]$/, "", $2); print $1"@"$2 }'`; do echo "$p"; meteor show "$p" | grep -E '^  [a-z]'; echo; done

**With that said**, `.meteor/versions` is a reasonable list of ALL packages, and running `meteor show <package-name>` is probably much more effective than running that command for every dependency in your entire project.

Also, I don't think the addition of a `--tree` argument to `meteor list` would be a terribly hard contribution.  This seems much more reasonable (and appreciated by the community) than making a third-party package which will take considerable work to ensure the constraints are respected in the same way that Meteor does.  The code is likely [somewhere in here](https://github.com/meteor/meteor/blob/97226b511d41e2832219cc287e2ee3bf0e895fea/tools/cli/commands-packages.js) and more information is likely available via [`getConstraints` in `project-context.js`](https://github.com/meteor/meteor/blob/devel/tools/project-context.js#L47-L48) and [`packageMapFile`](https://github.com/meteor/meteor/blob/devel/tools/project-context.js#L331-L334) – though my specific pointers may be a bit off.  Anyone want to make an issue/PR for that?

**Comment by aedm** [comment]
I wrote a script that displays dependency relations between Meteor packages, like so:

```
blaze@2.3.1
  depends on:
   - check
   - htmljs
   - jquery
   - observe-sequence
   - ordered-dict
   - reactive-var
   - tracker
   - underscore
  depended by:
   - accounts-base
   - spacebars
   - ui
```

Here it is: https://gist.github.com/aedm/979cf82ecaedc9bffa3d9872fc75d1b9

Be aware, it's insanely slow.

**Comment by abernix** [comment]
Thanks for offering that script, though I'd really suggest that the next person who wants to have this functionality help us build it into Meteor as opposed to building an external script (which is destined to be slow by the very fact that it is external to Meteor) by submitting a pull request.  I've made beginners suggestions above! 😬

The existing feature request for this seems to be at https://github.com/meteor/meteor/issues/4387.  I've marked it `pull-requests-encouraged`.
